File: library/src/test/java/io/github/dmitrikudrenko/logger2/LoggerPublishTest.java
package io.github.dmitrikudrenko.logger2;

import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.logging.Handler;
import java.util.logging.LogManager;
import java.util.logging.LogRecord;
import java.util.logging.Logger;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.*;

@RunWith(PowerMockRunner.class)
@PrepareForTest({Log.class})
public class LoggerPublishTest {
    private static final String MOCK_TAG = "tag";
    private static final String MOCK_MESSAGE = "message";

    private static Logger logger;
    private static Handler handler;

    @BeforeClass
    public static void setUp() {
        logger = mock(Logger.class);
        handler = mock(Handler.class);

        PowerMockito.mockStatic(LogManager.class);
        LogManager logManager = mock(LogManager.class);
        when(LogManager.getLogManager()).thenReturn(logManager);
        when(logManager.getLogger(any())).thenReturn(logger);
    }

    @Test
    public void shouldPublishRecord() {
        Log.addHandler(handler);
        Log.i(MOCK_TAG, MOCK_MESSAGE);
        verify(handler).publish(any(LogRecord.class));
    }
}

File: library/src/test/kotlin/io/github/dmitrikudrenko/logger2/LoggerPublishTest.kt
package io.github.dmitrikudrenko.logger2

import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Matchers.argThat
import org.mockito.Mockito.verify
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.util.logging.Handler
import java.util.logging.Level

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
class LoggerPublishTest : BasePublishTest() {

    @Test
    fun `should verbose record invoked`() {
        Log.v(MOCK_TAG, MOCK_MESSAGE)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Log.VERBOSE, MOCK_TAG, MOCK_MESSAGE)))
    }

    @Test
    fun `should verbose record with exception invoked`() {
        Log.v(MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Log.VERBOSE, MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)))
    }

    @Test
    fun `should debug record invoked`() {
        Log.d(MOCK_TAG, MOCK_MESSAGE)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Log.DEBUG, MOCK_TAG, MOCK_MESSAGE)))
    }

    @Test
    fun `should debug record with exception invoked`() {
        Log.d(MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Log.DEBUG, MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)))
    }

    @Test
    fun `should warning record invoked`() {
        Log.w(MOCK_TAG, MOCK_MESSAGE)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.WARNING, MOCK_TAG, MOCK_MESSAGE)))
    }

    @Test
    fun `should warning record with exception invoked`() {
        Log.w(MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.WARNING, MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)))
    }

    @Test
    fun `should info record invoked`() {
        Log.i(MOCK_TAG, MOCK_MESSAGE)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.INFO, MOCK_TAG, MOCK_MESSAGE)))
    }

    @Test
    fun `should info record with exception invoked`() {
        Log.i(MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.INFO, MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)))
    }

    @Test
    fun `should error record invoked`() {
        Log.e(MOCK_TAG, MOCK_MESSAGE)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.SEVERE, MOCK_TAG, MOCK_MESSAGE)))
    }

    @Test
    fun `should error record with exception invoked`() {
        Log.e(MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.SEVERE, MOCK_TAG, MOCK_MESSAGE, MOCK_EXCEPTION)))
    }

    @Test
    fun `should info record invoked if LogEvent sent`() {
        Log.event(MOCK_EVENT)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.INFO, MOCK_EVENT.javaClass.simpleName, MOCK_EVENT_VALUE)))
    }

    @Test
    fun `should info record with tag invoked if LogEvent sent`() {
        Log.event(MOCK_EVENT, MOCK_TAG)
        verify<Handler>(handler)
                .publish(argThat(LogRecordMatcher(Level.INFO, MOCK_TAG, MOCK_EVENT_VALUE)))
    }
}
